pkg/sql/resolver.go

```diff
@@ -299,69 +299,61 @@ func (p *planner) IsTypeVisible(
 	return false, true, nil
 }
 
-// HasPrivilege is part of the tree.EvalDatabase interface.
-func (p *planner) HasPrivilege(
+// HasAnyPrivilege is part of the tree.EvalDatabase interface.
+func (p *planner) HasAnyPrivilege(
 	ctx context.Context,
 	specifier tree.HasPrivilegeSpecifier,
 	user security.SQLUsername,
-	priv privilege.Privilege,
-) (bool, error) {
+	privs []privilege.Privilege,
+) (tree.HasAnyPrivilegeResult, error) {
 	desc, err := p.ResolveDescriptorForPrivilegeSpecifier(
 		ctx,
 		specifier,
 	)
 	if err != nil {
-		return false, err
+		return tree.HasNoPrivilege, err
+	}
+	if desc == nil {
+		return tree.ObjectNotFound, nil
 	}
 
-	// hasPrivilegeFunc checks whether any role has the given privilege.
-	hasPrivilegeFunc := func(priv privilege.Privilege) (bool, error) {
-		err := p.CheckPrivilegeForUser(ctx, desc, priv.Kind, user)
-		if err != nil {
+	for _, priv := range privs {
+		// RULE was only added for compatibility with Postgres, and Postgres
+		// never allows RULE to be granted, even if the user has ALL privileges.
+		// See https://www.postgresql.org/docs/8.1/sql-grant.html
+		// and https://www.postgresql.org/docs/release/8.2.0/.
+		if priv.Kind == privilege.RULE {
+			continue
+		}
+
+		if err := p.CheckPrivilegeForUser(ctx, desc, priv.Kind, user); err != nil {
 			if pgerror.GetPGCode(err) == pgcode.InsufficientPrivilege {
-				return false, nil
+				continue
 			}
-			return false, err
+			return tree.HasNoPrivilege, err
 		}
 
 		if priv.GrantOption {
 			if !p.ExecCfg().Settings.Version.IsActive(ctx, clusterversion.ValidateGrantOption) {
-				err := p.CheckPrivilegeForUser(ctx, desc, privilege.GRANT, user)
-				if err != nil {
+				if err := p.CheckPrivilegeForUser(ctx, desc, privilege.GRANT, user); err != nil {
 					if pgerror.GetPGCode(err) == pgcode.InsufficientPrivilege {
-						return false, nil
+						continue
 					}
-					return false, err
+					return tree.HasNoPrivilege, err
 				}
 			} else {
-				err := p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, user, true /* isGrant */)
-				if err != nil {
+				if err := p.CheckGrantOptionsForUser(ctx, desc, []privilege.Kind{priv.Kind}, user, true /* isGrant */); err != nil {
 					if pgerror.GetPGCode(err) == pgcode.WarningPrivilegeNotGranted {
-						return false, nil
+						continue
 					}
-					return false, err
+					return tree.HasNoPrivilege, err
 				}
 			}
 		}
-
-		return true, nil
+		return tree.HasPrivilege, nil
 	}
 
-	if priv.Kind == privilege.RULE {
-		// RULE was only added for compatibility with Postgres, and Postgres
-		// never allows RULE to be granted, even if the user has ALL privileges.
-		// See https://www.postgresql.org/docs/8.1/sql-grant.html
-		// and https://www.postgresql.org/docs/release/8.2.0/.
-		return false, nil
-	}
-	hasPrivilege, err := hasPrivilegeFunc(privilege.Privilege{Kind: privilege.ALL})
```

Comment:
`privilege.ALL` does not need special handling here because `CheckPrivilegeForUser` and `CheckGrantOptionsForUser` handle that case internally.